### PR TITLE
Resizing event (v2)

### PR DESCRIPTION
**Pull request by nmschulte**
This is a rebased version of #292, for easy inclusion into master.  Thanks @emitron for the work.  @jackmoore, is this PR alright?  As it stands, it causes the `autosize:resized` event to be fired, even if there was no resize; this allows hooking before a resize event and effecting behavior from that in the listener/client code.  I wouldn't be opposed to e.g. a separate event for this, and only issuing it when an `autosize:resized` event would be issued.

From the other PR, this is what I said:

> I am trying to solve a similar / the same problem: need to conditionally scroll a different element whose height the autosizing might have affected, depending on e.g. the previous height (and/or scroll). an autosize:beforeResized event, or similar. It's not clear how this change works given the description, but I'm in the middle of rebasing the code for merging to master.

### CODE REVIEW

**Comment by jackmoore** [comment]
Help me understand what this would be used for, I don't want to add a new event unless there is a good use case for it.

**Comment by nmschulte** [comment]
Just so it's clear; this is just a re-base (to fix merge conflicts) of the other PR.  The use case is the same, IIUC.

@jackmoore you said, in #292:
> I think the last release may fix the issue you were working around by adding this event and saving your UI state. It preserves the scroll position on ancestor elements, which is what it sounds like you are manually controlling for.

This doesn't actually resolve my issue, and maybe not @emirotin's either.  In my case, the element that needs dealt w/ is not a parent/ancestor, it's a sibling.

The `autosize:resizing` event acts as a cue to the client so it can capture any state it needs so it can do the logic after receiving an `autosize:resized` event.  To be honest, I would prefer to see the `autosize:resizing` event guarded and only issued if an `autosize:resized` event will follow (read: do the calc; if necessary issue the `resizing`  && then issue the `resized` events.

**Comment by jackmoore** [comment]
> I need to maintain a scroll position on one of its siblings

That makes sense to me.

> I would prefer to see the autosize:resizing event guarded and only issued if an autosize:resized event will follow

This would be ideal, but there is no way of doing that with Autosize's current approach to determining height. The act of determining if the textarea height needs to be changed, changes the textarea height temporarily.  I'm assuming that temporary change is what you need to preempt.

Autosize calculates the size each time the input event fires, so you could add your own event handler for that before assigning autosize, instead of relying on a custom event.  For example:

```
function doSomething() {};

var ta = document.querySelector('textarea');
ta.addEventListener('input',doSomething);
ta.addEventListener('autosize:update', doSomething);
autosize(ta);
```

**Comment by nmschulte** [comment]
I'm not certain what the expected browser behavior is w/ regard to scrolling: if the element is resized, is the browser supposed to maintain the scroll "position" (the height-agnostic value; e.g. a percentage) or the absolute `scrollTop` value, or ???

I don't need to preempt the temporary height change; the case I care about is the one where the height actually _does_ change, and the temporary height change to check for a height change is not causing a problem.

(I'm not certain what the expected browser behavior is w/ regard to scrolling: if the element is resized, is the browser supposed to maintain the scroll "position" (the height-agnostic value; e.g. a percentage) or the absolute `scrollTop` value, or ???

It may be that the temporary re-size is what's keeping the sibling from maintaining its scroll position as I would like, but that would imply a bug somewhere in the browser and I doubt that.)

You're right about your proposed solution, but it leaves maintenance holes.  Does autosize _only_ listen for the `input` event?  I know that, e.g., to support non-modern browsers, sometimes there are other events and dances of concern, and I don't want to have to replicate those concerns.

File: src/autosize.js [hunk 1]
```diff
@@ -142,9 +152,8 @@ function assign(ta) {
 
 		if (cachedHeight !== computedHeight) {
 			cachedHeight = computedHeight;
-			const evt = createEvent('autosize:resized');
-			ta.dispatchEvent(evt);
 		}
+        dispatchEvent(ta, 'autosize:resized');
```
Comment (emirotin) [inline src/autosize.js]: Looks like I've messed with tabs and spaces here

**Comment by emirotin** [comment]
> This would be ideal, but there is no way of doing that with Autosize's current approach to determining height. 

I don't remember much of the details since I worked on this long time ago, and am using my fork since then without any changes.
But if that would be possible I know I would fire the event only when necessary. That's why it's implemented as it is.

> I'm not certain what the expected browser behavior is w/ regard to scrolling: if the element is resized, is the browser supposed to maintain the scroll "position" (the height-agnostic value; e.g. a percentage) or the absolute scrollTop value, or ???

The browser maintains the absolute scrollTop according to my observations.

A chat app is exactly my case. As the input area grows the messages list shrinks with scroll getting detached from its bottom.

**Pull request closed without merging**